consistent-sampling/src/main/java/io/opentelemetry/contrib/sampler/consistent2/ConsistentSampler.java

```diff
@@ -0,0 +1,319 @@
+   * called for each sampler individually. Also, the combined sampler is more efficient than
+   * evaluating the two samplers individually and combining both results afterwards.
+   *
+   * @param otherConsistentSampler the other consistent sampler
+   * @return the composed consistent sampler
+   */
+  public ConsistentSampler or(ConsistentSampler otherConsistentSampler) {
+    if (otherConsistentSampler == this) {
+      return this;
+    }
+    return new ConsistentComposedOrSampler(
+        this, otherConsistentSampler, RandomValueGenerators.getDefault());
+  }
+
+  private final RandomValueGenerator randomValueGenerator;
+
+  protected ConsistentSampler(RandomValueGenerator randomValueGenerator) {
+    this.randomValueGenerator = requireNonNull(randomValueGenerator);
+  }
+
+  @Override
+  public final SamplingResult shouldSample(
+      Context parentContext,
+      String traceId,
+      String name,
+      SpanKind spanKind,
+      Attributes attributes,
+      List<LinkData> parentLinks) {
+
+    Span parentSpan = Span.fromContext(parentContext);
+    SpanContext parentSpanContext = parentSpan.getSpanContext();
+    boolean isRoot = !parentSpanContext.isValid();
+    boolean isParentSampled = parentSpanContext.isSampled();
+
+    boolean isRandomTraceIdFlagSet = false; // TODO in future get the random trace ID flag, compare
+    // https://www.w3.org/TR/trace-context-2/#random-trace-id-flag
+
+    TraceState parentTraceState = parentSpanContext.getTraceState();
+    String otelTraceStateString = parentTraceState.get(OtelTraceState.TRACE_STATE_KEY);
+    OtelTraceState otelTraceState = OtelTraceState.parse(otelTraceStateString);
+
+    long randomValue;
+    if (otelTraceState.hasValidRandomValue()) {
+      randomValue = otelTraceState.getRandomValue();
+    } else if (isRandomTraceIdFlagSet) {
+      randomValue = OtelTraceState.parseHex(traceId, 18, 14, getInvalidRandomValue());
+    } else {
+      randomValue = randomValueGenerator.generate(traceId);
+      otelTraceState.invalidateThreshold();
+      otelTraceState.setRandomValue(randomValue);
+    }
+
+    long parentThreshold;
+    if (otelTraceState.hasValidThreshold()) {
+      long threshold = otelTraceState.getThreshold();
+      if (((randomValue < threshold) == isParentSampled) || threshold == 0) {
+        parentThreshold = threshold;
+      } else {
+        parentThreshold = getInvalidThreshold();
+      }
+    } else if (isParentSampled) {
+      parentThreshold = getMaxThreshold();
+    } else {
+      parentThreshold = 0;
+    }
+
+    // determine new threshold that is used for the sampling decision
+    long threshold = getThreshold(parentThreshold, isRoot);
+
+    // determine sampling decision
+    boolean isSampled;
+    if (isValidThreshold(threshold)) {
+      isSampled = (randomValue < threshold);
+      if (0 < threshold && threshold < getMaxThreshold()) {
+        otelTraceState.setThreshold(threshold);
```

Comment:
Setting this threshold in the trace state regardless of the `isSampled` value may lead to violation of the invariant `(RV < T) == sampled || T == 0`. That may be okey, but was it intended?

Reply:
The sampled flag is not part of the trace state. However, `isSampled` is properly set and later used for the sampling decision. If I'm not missing anything, this should not break the invariant.

Reply:
Yes, you are right. I don't remember what I was thinking when I wrote my comment.